sycl/plugins/level_zero/pi_level_zero.cpp

```diff
@@ -791,14 +782,47 @@ static const size_t ImmCmdListsEventCleanupThreshold = [] {
 }();
 
 // Whether immediate commandlists will be used for kernel launches and copies.
-// The default is standard commandlists. Setting a value >=1 specifies use of
+// The default is standard commandlists. Setting 1 or 2 specifies use of
 // immediate commandlists. Note: when immediate commandlists are used then
 // device-only events must be either AllHostVisible or OnDemandHostVisibleProxy.
 // (See env var SYCL_PI_LEVEL_ZERO_DEVICE_SCOPE_EVENTS).
-bool _pi_device::useImmediateCommandLists() {
+
+enum ImmCmdlistMode {
+  // Immediate commandlists are not used.
+  NotUsed,
+  // One set of compute and copy immediate commandlists per queue.
+  PerQueue,
+  // One set of compute and copy immediate commandlists per host thread that
+  // accesses the queue.
+  PerThreadPerQueue
+};
+
+// Get value of immediate commandlists env var setting or -1 if unset
+// A value of 1 specifies a single set of imm cmdlists per queue.
+// A value of 2 specifies a separate set of imm cmdlists per thread per queue.
+static const int ImmediateCommandlistsSetting = [] {
+  const char *ImmediateCommandlistsSettingStr =
+      std::getenv("SYCL_PI_LEVEL_ZERO_USE_IMMEDIATE_COMMANDLISTS");
+  if (!ImmediateCommandlistsSettingStr)
+    return -1;
+  return std::stoi(ImmediateCommandlistsSettingStr);
+}();
+
+int _pi_device::useImmediateCommandLists() {
+  // If immediate commandlist setting is not explicitly set, then use the device
+  // default.
   if (ImmediateCommandlistsSetting == -1)
     return ImmCommandListsPreferred;
-  return ImmediateCommandlistsSetting;
+  switch (ImmediateCommandlistsSetting) {
+  case 0:
+    return NotUsed;
+  case 1:
+    return PerQueue;
+  case 2:
+    return PerThreadPerQueue;
+  default:
+    return NotUsed;
+  }
```

Comment:
please combine these into a single function reading the setting and returning enum

Reply:
Ok, done.